.github/workflows/pipeline.yml

```diff
@@ -60,12 +60,30 @@ jobs:
         with:
           path: sytest
 
+      # TODO the shell script below is nicked from complement. We use this pattern
+      # in a few places. Can we make this an Action so it's easier to reuse?
       - name: Fetch corresponding synapse branch
         shell: bash
         run: |
-          BRANCH=${GITHUB_REF#refs/heads/}
-          (wget -O - https://github.com/matrix-org/synapse/archive/$BRANCH.tar.gz || wget -O - https://github.com/matrix-org/synapse/archive/develop.tar.gz) \
-            | tar -xz --strip-components=1 -C /src/
+          # Attempt to use the version of synapse which best matches the current
+          # build. Depending on whether this is a PR or release, etc. we need to
+          # use different fallbacks.
+          #
+          # 1. First check if there's a similarly named branch (GITHUB_HEAD_REF
+          #    for pull requests, otherwise GITHUB_REF).
+          # 2. Attempt to use the base branch, e.g. when merging into release-vX.Y
+          #    (GITHUB_BASE_REF for pull requests).
+          # 3. Use the default synapse branch ("develop").
+          for BRANCH_NAME in "$GITHUB_HEAD_REF" "$GITHUB_BASE_REF" "${GITHUB_REF#refs/heads/}" "develop"; do
```

Comment:
Since we had an issue with `"${GITHUB_REF#refs/heads/}"` due to PR numbers, is it even worth including it on this list?

Reply:
GITHUB_HEAD_REF and GITHUB_BASE_REF only exist when the action is triggered by a pull request. https://docs.github.com/en/actions/reference/environment-variables#default-environment-variables. This action runs on specific branches too, outside of the context of pull requests:

```
name: Run sytest
on:
  push:
    branches: ["develop", "release-*"]
  pull_request
```

(but note that's only there because I copied it from synapse's CI file)

Assuming it's meaningful to run sytest on develop and release branches, I think we do still want to read GITHUB_REF.

Reply:
Aha, I see. So in the case of a pull request, it's unlikely that we'll end up using `"${GITHUB_REF#refs/heads/}"`. But it's useful on `develop` / `release-*`.

Yes, I think it's worthwhile to run Sytest on those branches.